### PR TITLE
Relocate 4xx/5xx exception in downloadFile

### PR DESCRIPTION
**Pull request by vercingetorx**
Move 4xx/5xx exception to before disk i/o. As it stands an empty file is created on http error 4xx/5xx.

### CODE REVIEW

**Comment by timotheecour** [comment]
* i closed/re-opened so that the PR that fixed CI on i386 gets merged in
* i don't understand why you had to open 3 different PR's for that instead of reusing one via rebase + force push (https://github.com/nim-lang/Nim/pull/17328, https://github.com/nim-lang/Nim/pull/17329)

**Comment by vercingetorx** [comment]
Thanks. I just felt it was better to start from scratch and redo the pull  request since I messed up the others. My mistake.

**Review by timotheecour — APPROVED** [review]

File: lib/pure/httpclient.nim [hunk 1]
```diff
@@ -1219,21 +1219,24 @@ proc downloadFile*(client: HttpClient, url: Uri | string, filename: string) =
   defer:
     client.getBody = true
   let resp = client.get(url)
+  
+  if resp.code.is4xx or resp.code.is5xx:
+    raise newException(HttpRequestError, resp.status)
 
   client.bodyStream = newFileStream(filename, fmWrite)
   if client.bodyStream.isNil:
     fileError("Unable to open file")
```
Comment (timotheecour) [inline lib/pure/httpclient.nim:1228]: future PR's welcome to improve error messages, eg:
`fileError("Unable to open file")`
=>
`fileError("Unable to open file: '%1'" %  filename)` or similar

**Comment by timotheecour** [comment]
no worries, I see this is your 1st PR in nim repo, looking forward to more, thanks!

**Pull request merged**